src/core/dense_set.cc

```diff
@@ -0,0 +1,424 @@
+// Copyright 2022, Roman Gershman.  All rights reserved.
+// See LICENSE for licensing terms.
+//
+
+#include "core/dense_set.h"
+
+#include <absl/numeric/bits.h>
+
+#include <cstddef>
+#include <cstdint>
+#include <stack>
+#include <type_traits>
+#include <vector>
+
+#include "glog/logging.h"
+
+extern "C" {
+#include "redis/zmalloc.h"
+}
+
+namespace dfly {
+using namespace std;
+
+constexpr size_t kMinSizeShift = 2;
+constexpr size_t kMinSize = 1 << kMinSizeShift;
+constexpr bool kAllowDisplacements = true;
+
+DenseSet::DenseSet(pmr::memory_resource* mr) : entries_(mr) {
+}
+
+DenseSet::~DenseSet() {
+  Clear();
+}
+
+size_t DenseSet::PushFront(DenseSet::ChainVectorIterator it, void* data) {
+  // if this is an empty list assign the value to the empty placeholder pointer
+  if (it->IsEmpty()) {
+    it->SetObject(data);
+    return ObjectAllocSize(data);
+  }
+
+  // otherwise make a new link and connect it to the front of the list
+  LinkAllocator la(mr());
```

Comment:
nit: maybe introduce an inline helper function that receives a memory resource, data and DensPtr `next` and returns `DenseLinkKey*` ? (i.e to remove the duplication in the next function)